lambdajs/src/SharedAbstract.hs

```diff
@@ -0,0 +1,262 @@
+    deleteField f1 = proc (_, _) -> returnA -< singleton TTop
+    -- operator/delta function
+    evalOp = proc (op, vals) -> do
+        case length vals of
+            1 -> do
+                t <- Control.Arrow.Utils.map typeEvalUnOp_ -< zip (repeat op) (Data.Set.toList $ head vals)
+                returnA -< Data.Set.fromList t
+            2 -> do
+                let
+                    v1 = Data.Set.toList (head vals)
+                    v2 = Data.Set.toList (head $ tail vals)
+                case length v1 == 1 && length v2 == 1 of
+                    True -> do
+                        t <- typeEvalBinOp_ -< (op, head v1, head v2)
+                        returnA -< Data.Set.fromList [t]
+                    False -> fail -< "Error: Binary op with set of type params not supported"
+    -- environment ops
+    lookup = proc id -> do
+        Control.Arrow.Environment.lookup pi1 Control.Category.id -< (id, Data.Set.singleton TUndefined)
+    apply f1 = proc (lambdas, args) -> do
+        ts <- Control.Arrow.Utils.map (proc (lambda, args) -> case lambda of
+            TLambda names body closureEnv -> do
+                case length names == length args of
+                    False -> fail -< "Error: lambda must be applied with same amount of args as params"
+                    True -> do
+                        newBindings <- arr $ uncurry zip -< (names, args)
+                        bindingEnv <- bindings -< (newBindings, closureEnv)
+                        outsideEnv <- getEnv -< ()
+                        finalEnv <- bindings -< (Data.Abstract.Environment.toList bindingEnv, outsideEnv)
+                        localEnv f1 -< (finalEnv, body)
+            _ -> returnA -< Data.Set.fromList [TObject [("0", singleton TString), ("length", singleton TNumber), ("$isArgs", singleton TBool)]]) -< zip (Data.Set.toList lambdas) (repeat args)
+        returnA -< Prelude.foldr Data.Set.union (Data.Set.empty) ts
+    -- store ops
+    set = proc (loc, val) -> do
+        case Data.Set.toList loc of
+            [TRef l] -> do
+                Control.Arrow.Utils.map write -< zip (Data.Set.toList l) (repeat val)
+                returnA -< ()
+            _ -> fail -< "Error: ESetRef lhs must be location"
+    new = proc (val) -> do
+        loc <- fresh >>> (arr (:[])) >>> (arr Data.Set.fromList) -< ()
+        set -< (Data.Set.fromList [TRef loc], val)
+        returnA -< Data.Set.fromList [TRef loc]
+    get = proc (loc) -> do
+        case Data.Set.toList loc of
+            [TRef l] -> do
+                vals <- Control.Arrow.Utils.map (read pi1 Control.Category.id) -< zip (Data.Set.toList l) (repeat (Data.Set.singleton TUndefined))
+                returnA -< foldr1 (\x y -> x ⊔ y) vals
+            _ -> returnA -< (Data.Set.fromList [TThrown (Data.Set.fromList [TObject []])])
+    -- control flow
+    if_ f1 f2 = proc (cond, thenBranch, elseBranch) -> do
+        case Data.Set.toList cond of
+            [TBool] -> do
```

Comment:
Use the least upper bound operator on the arrow computations.
```
(f1 -< thenBranch) ⊔ (f2 -< elseBranch)
```
or
```
joined f1 f2 -< (thenBranch,elseBranch)
```

In general, it is better to join computations that just the values, otherwise you may not get what you want. For example, if the evaluation of the `thenBranch` fails, you might not even get to the evaluation of the `elseBranch`. Joining on computations takes care of that.